### DISCUSSION TITLE
Implementation of HCD functions for STM32

### DISCUSSION THREAD

**Discussion by ghost** [hathach/tinyusb#857] [Q&A] [answered]
When trying to embed the library in a third-party project, following the instructions, or to build an example from the library for STM32 in host mode, the same errors of lack of implementation (definition) of some functions related to HCD appear.

> $ make
  LD    test.elf
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `_osal_q_lock':
tinyusb/src/osal/osal_none.h:143: undefined reference to `hcd_int_disable'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `_osal_q_unlock':
tinyusb/src/osal/osal_none.h:157: undefined reference to `hcd_int_enable'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `hcd_frame_number':
tinyusb/src/host/hcd.h:112: undefined reference to `hcd_uframe_number'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `tuh_init':
usbh.c:202: undefined reference to `hcd_init'
arm-none-eabi/bin/ld.exe: usbh.c:203: undefined reference to `hcd_int_enable'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `usbh_edpt_xfer':
usbh.c:263: undefined reference to `hcd_edpt_xfer'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `usbh_pipe_control_open':
usbh.c:278: undefined reference to `hcd_edpt_open'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `usbh_edpt_open':
usbh.c:283: undefined reference to `hcd_edpt_open'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `usbh_device_unplugged':
usbh.c:396: undefined reference to `hcd_device_close'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `enum_new_device':
usbh.c:663: undefined reference to `hcd_port_connect_status'
arm-none-eabi/bin/ld.exe: usbh.c:665: undefined reference to `hcd_port_speed_get'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `enum_get_addr0_device_desc_complete':
usbh.c:731: undefined reference to `hcd_port_reset'
arm-none-eabi/bin/ld.exe: ../bin/./usbh.o: in function `enum_set_address_complete':
usbh.c:765: undefined reference to `hcd_device_close'
arm-none-eabi/bin/ld.exe: ../bin/./usbh_control.o: in function `tuh_control_xfer':
usbh_control.c:76: undefined reference to `hcd_setup_send'
arm-none-eabi/bin/ld.exe: ../bin/./usbh_control.o: in function `usbh_control_xfer_cb':
usbh_control.c:111: undefined reference to `hcd_edpt_xfer'
arm-none-eabi/bin/ld.exe: usbh_control.c:126: undefined reference to `hcd_edpt_xfer'
collect2.exe: error: ld returned 1 exit status
make: *** [../rules.mk:145: test.elf] Error 1

Do I understand correctly that the implementation of these platform-dependent functions falls on the shoulders of the developer using the library?

If I'm wrong, please tell me where to find the implementation for STM32F407VG, because definitions for only some devices are available in the source files.

**Comment by hathach** [comment, marked as answer]
it is not yet implemented. the host stack only support some lpc, imxrt (ehci, ohci) and rp2040

**Reply by ghost** [reply]
How soon do you plan to add? Are there any relatively accurate predictions?
